nav2_bringup/bringup/params/nav2_params.yaml

```diff
@@ -113,7 +113,7 @@ local_costmap:
       obstacle_layer:
         enabled: True
         scan:
-          topic: /scan
+          topic: scan
```

Comment:
This shouldn't work as I understand it. Since the costmaps are in a namespace, without the leading /, the topics we subscribe to will also be in the namespace. That is, the global and local costmaps would end up subsribing to `global_costmap/scan` and `local_costmap/scan` respectively.

The same problem would exist for the rest of the topics in the costmap nodes.

Am I missing something?

Reply:
It wont necessarily go into the namespace unless when the node for the process is created was given a namespace to use ei `("node_name", "ns", false)`, if its `("node_name", "", false)` itll use the global namespace for topics. But I can verify that this is correct but I'm pretty sure it is if its passing CI

Reply:
I was just recently made aware that we don't have any checks on obstacle detection. Since we are only using static maps with no dynamic obstacles, we still pass the system test without `scan` working.

Reply:
Well they dont have to be dynamic, but I do suppose they would have to not be in the map, or the global planner doesn't use the static layer.

I'll look into it. If a parameter file needs to have an absolute path that's fine

Reply:
I tested the change and can confirm this causes the robot to bump into objects not considered on the map since it's now looking for the topic in `/robot1/local_costmap/scan`.

Two options:
1. Keep the fully qualified name for the topic and the remapping for `scan`.
2. Remove the `/` and remapping, and somehow pass the "parent node" namespace to the obstacle layer so it subscribes to "/parent_node_ns/scan". In this case the parent node is `/robot1/local_costmap` and it's ns is '/robot1/`. We're doing something similar for the map here: https://github.com/ros-planning/navigation2/blob/ada1b648eb2e6e029a23eb5e4af956bfa7387cca/nav2_costmap_2d/src/costmap_2d_ros.cpp#L91 For example, we could pass it by defining a new parameter somewhere in here: https://github.com/ros-planning/navigation2/blob/ada1b648eb2e6e029a23eb5e4af956bfa7387cca/nav2_costmap_2d/src/costmap_2d_ros.cpp#L93
like: ```cpp
declare_parameter("parent_node_namespace", rclcpp::ParameterValue(parent_namespace));
``` and pulling it and applying it to the scan topic here: https://github.com/ros-planning/navigation2/blob/ada1b648eb2e6e029a23eb5e4af956bfa7387cca/nav2_costmap_2d/plugins/obstacle_layer.cpp#L130

Reply:
OK,  sorry I didnt get to testing this today. I wouldn't want to change the code to support, can't we just have in the param file `/scan`? It _is_ a configuration file, its meant to have specific configurations for a specific case. No remapping required (just 2 configurations, which seems reasonable to me) ((or the yaml loader have an option to overload a param))

Reply:
If I understand correctly, you're proposing a third option: set the correct fully-qualified name of the topic on the params file, for example `/robot_ns/scan`. Of the two options you mention, I'd prefer that the param's default value is replaced at launch instead of having multiple files.

I'd be ok with options 2 and 3 -- leaning more towards 3.

Reply:
>  I'd prefer that the param's default value is replaced at launch instead of having multiple files

I think the problem with that is that doesn't scale. Today its `scan` but there's no bound on the arbitrary parameters we've want to push into another namespace, but ignoring others. 

Multiple configuration files makes alot of sense to me, unless we applied the namespacing to every single parameter in the parameter file so there werent special cases

Reply:
I'm completely convinced we should consider placing the stack on a namespace a particular use case. Why not have it do the right thing by default?

I think part of the problem is expressing the design intent on the params file. For example, in the case of `/scan`, it was previously clear. However, now to support namespacing, by default it should probably be `../scan` or `<nav_ns>/scan` where the tokens get replaced at launch. This would work fine on the single or multi-robot cases. Also, if the user sets it to `scan`, I think the right way to interpret this is that the topic should be  `local_costmap/scan`.

This is applicable to any case where the user wants the stack namespaced even in the single robot scenario, e.g. `navigation/...`, `safety/..`, `manipulation/...`,

Reply:
> I'm not completely convinced we should consider placing the stack on a namespace a particular use case

I would actually disagree that I don't think it matters. If we had a default namespace of `/nav2/...` and all our stuff was inside of that, I'd argue that was *incredibly* well organized for userspace code. If someone looks at their `ros2 topic list` then they could instantly separate out their code from the metapackage code. 

The added benefit then is trivial additions for multiple robots, rather than `/nav2/..` you can replace with robot names `/freight/...` `/tally/...` `/locusbot/...` when you want to have multiple robots in concert. 

Param files stay the same, just all nodes and all things are given that header added

Reply:
WIth that said, that's a much larger change and I'm not trying to do that right now.

So what are your thoughts for the quickest reasonable option here? Just add the `/` back in front of scan in the params file? Then for multi-robot configs, we have an example with `/ns/scan` rather than `/scan`?

Reply:
That sounds reasonable.

@crdelsey do you have a preference?

Reply:
I think putting the stack in a `nav2/` namespace is the best solution in the long run. We also have talked about getting rid of the costmap as a separate node, and just passing in the parent node. That would also eliminate this problem.

I think both those actions are desirable, so, I'd recommend doing the expedient solution today and revisiting when either of the above improvements are done.

Reply:
>  We also have talked about getting rid of the costmap as a separate node, and just passing in the parent node. That would also eliminate this problem.

Well... that would make things super annoying to deal with if we remove the costmaps from the controller/planner servers themselves. But I dont see that happening anytime in the immediate future with our future list plans

Reply:
So... to recap, just add the `/` back in front of `/scan` in the params file?

Reply:
> that would make things super annoying to deal with if we remove the costmaps from the controller/planner servers themselves.

That's not what suggesting. Rather, when you construct the costmap_2d_ros object, you'd pass in the controller/planner server node. Costmap_2d_ros would add it's pubs/subs to that node instead of being its own node. With the exeception of possibly switching to multithreaded executors, nothing else changes from the server or plugin perspective.

Given that we use costmap as an object - ie we make method calls to it from planner/controller server rather than calling services and sending messages, that seems more consistent with other objects in the stack.

There are no firm plans to make that change, but it keeps coming up as "something we should do"

>  just add the / back in front of /scan in the params file

I think that's what I'm hearing. @orduno Do you agree?

Reply:
pinging @orduno

Reply:
Yes, have a robot params file with the topic as `/scan`, and as you suggested, a second multi-robot params file with `/<robotID>/scan`. 

I'd prefer not having duplicated params file, but if you prefer this option than to keep the remapping on the topics, then I'm ok.

I can implement the `PushRemapping` that would resolve this problem.